src/Features/Core/Portable/IntroduceVariable/AbstractIntroduceParameterService.cs

```diff
@@ -0,0 +1,951 @@
+                ImmutableArray<IParameterSymbol> validParameters, string newMethodIdentifier, int insertionIndex, CancellationToken cancellationToken)
+            {
+                var invocationArguments = syntaxFacts.GetArgumentsOfArgumentList(argumentListSyntax);
+                var parameterToArgumentMap = MapParameterToArgumentsAtInvocation(semanticFacts, invocationArguments, invocationSemanticModel, cancellationToken);
+                var currentInvocationArguments = syntaxFacts.GetArgumentsOfArgumentList(currentArgumentListSyntax);
+                var requiredArguments = new List<SyntaxNode>();
+
+                foreach (var parameterSymbol in validParameters)
+                {
+                    if (parameterToArgumentMap.TryGetValue(parameterSymbol, out var index))
+                    {
+                        requiredArguments.Add(currentInvocationArguments[index]);
+                    }
+                }
+
+                var methodName = generator.IdentifierName(newMethodIdentifier);
+                var newMethodInvocation = generator.InvocationExpression(methodName, requiredArguments);
+                var allArguments = AddArgumentToArgumentList(currentInvocationArguments, generator, newMethodInvocation, insertionIndex);
+                return syntaxFacts.UpdateArgumentListSyntax(currentArgumentListSyntax, allArguments);
+            }
+        }
+
+        /// <summary>
+        /// Generates a method declaration containing a return expression of the highlighted expression.
+        /// Example:
+        /// public void M(int x, int y)
+        /// {
+        ///     int f = [|x * y|];
+        /// }
+        /// 
+        /// ---------------------------------------------------->
+        /// 
+        /// public int GetF(int x, int y)
+        /// {
+        ///     return x * y;
+        /// }
+        /// 
+        /// public void M(int x, int y)
+        /// {
+        ///     int f = x * y;
+        /// }
+        /// </summary>
+        private static async Task<SyntaxNode> ExtractMethodAsync(Document document, TExpressionSyntax expression,
+            IMethodSymbol methodSymbol, ImmutableArray<IParameterSymbol> validParameters, string newMethodIdentifier, SyntaxGenerator generator, CancellationToken cancellationToken)
+        {
+            // Remove trivia so the expression is in a single line and does not affect the spacing of the following line
+            var newStatements = ImmutableArray.CreateBuilder<SyntaxNode>();
+            newStatements.Add(generator.ReturnStatement(expression.WithoutTrivia()));
+            var semanticModel = await document.GetRequiredSemanticModelAsync(cancellationToken).ConfigureAwait(false);
+            var typeSymbol = semanticModel.GetTypeInfo(expression, cancellationToken).ConvertedType ?? semanticModel.Compilation.ObjectType;
```

Comment:
extract this repeated code into a helper

Reply:
done